Use replace when providing entire document; pass ObjectMapper to code…

alechenninger opened the pull request:
…c registry

Fixes #170
Fixes #169

Obviously, this class needs more tests... which unfortunately I don't have time to write right now :-)

alechenninger commented:
@jroper @benmccann Any thoughts on this? Currently using a fork until these issues are fixed

cfogrady commented:
I apologize. I'm in the middle of moving and house purchasing this month, but I'll try to take a look at this Monday.

alechenninger commented:
Ack, thanks for the update. Actually, I can relate, as I'm going through the exact same thing right now :-).

jroper commented on src/main/java/org/mongojack/JacksonCodecRegistry.java:
```diff
@@ -38,8 +42,8 @@ public JacksonCodecRegistry(ObjectMapper objectMapper, Class<?> view) {
         codecRegistry = MongoClient.getDefaultCodecRegistry();
     }
 
-    public JacksonCodecRegistry() {
```
Deprecate.

jroper commented on src/main/java/org/mongojack/JacksonMongoCollection.java:
```diff
@@ -352,24 +352,24 @@ public UpdateResult updateMany(DBQuery.Query query, DBUpdate.Builder update, boo
      * @throws MongoException
      *             If an error occurred
      */
-    public UpdateResult update(DBQuery.Query query, T object, boolean upsert, WriteConcern concern) throws MongoException, MongoWriteException,
```
Deprecate.

jroper commented on src/main/java/org/mongojack/JacksonMongoCollection.java:
```diff
@@ -382,14 +382,14 @@ public UpdateResult update(DBQuery.Query query, T object, boolean upsert, WriteC
      * @throws MongoException
      *             If an error occurred
      */
-    public UpdateResult update(Document query, T object, boolean upsert, WriteConcern concern) throws MongoException, MongoWriteException,
```
Deprecate.

jroper commented on src/main/java/org/mongojack/JacksonMongoCollection.java:
```diff
@@ -449,18 +449,18 @@ public UpdateResult update(DBQuery.Query query, DBUpdate.Builder update)
      * @throws MongoException
      *             If an error occurred
      */
-    public UpdateResult update(DBQuery.Query query, T object)
```
Deprecate.

jroper commented on src/main/java/org/mongojack/JacksonMongoCollection.java:
```diff
@@ -469,8 +469,8 @@ public UpdateResult update(DBQuery.Query query, T object)
      *             If an error occurred
      */
-    public UpdateResult updateById(Object _id, T object) throws MongoException, MongoWriteException, MongoWriteConcernException {
-        return update(createIdQuery(_id), object, false, null);
```
Deprecate.

benmccann commented:
@alechenninger do you plan to address the comments that James left? Also, this PR would need to be rebased

alechenninger commented:
@benmccann @jroper I have pushed some updates. Also, I'd like to elaborate on a few things based on James' comment:

> A few comments:
> 
> Please deprecate any methods you've deleted, rather than just deleting them, and include a comment on what the new method should be, so that people upgrading to this version know how to migrate and so that libraries that transitively depend on this don't just break.

I believe the update methods that take whole documents never worked, because they are using the wrong mongo client API which per javadoc, "update – a document describing the update, which may not be null. **The update to apply must include only update operators.**" However, this is passing a whole document (not update operators). The right mongo client API is `replaceOne`, so I took the opportunity to rename the methods to match.

I've added them back per request, but I just wanted to be clear about the reasoning there, because I didn't describe it in the original PR description (sorry about that) and in case you see that line of reasoning.

> It's been over 6 years since I wrote this, and probably 5 years since I've touched MongoDB, so I could be completely wrong here, but I think the intention for these methods was that if you just have a simple update where you're just updating one field, you could pass an object here that would set that field and leave everything else as null. The result would be a document that just contained that one field, and I thought MongoDB would treat that as simply setting that field, and leaving the rest of the object untouched. Is that right? If that's the case, then perhaps the old methods should be left as is, perhaps with comments updated to better explain how they function, and the replaceOne methods (which I don't think existed back then, I replicated most of the Java API so the fact that they aren't there indicates that they are new) should just be added.

AFAICT this class was added this year as part of 3.x. Maybe you're referring to another class?

Update does not work quite as described for this particular API for reasons described above. The other update APIs should work and remain untouched.

> Tests would be good - particularly at least one test to verify that my understanding of the old update behaviour was correct. From memory, writing tests was quite straight forward for this library.

Added some tests for the new APIs. FWIW, the tests do not run from IntelliJ, at least out of the box.

Thank you very much for the review! Apologies for the delay on updates.

benmccann commented on src/main/java/org/mongojack/JacksonMongoCollection.java:
```diff
@@ -334,54 +335,18 @@ public UpdateResult updateMany(DBQuery.Query query, DBUpdate.Builder update, boo
-            return mongoCollection.updateOne(serializeQuery(query), convertToDocument(object), new UpdateOptions().upsert(upsert));
-        }
-
+        return replaceOne(query, object, upsert, concern);
```
Let's not change the implementation of this method. It would be dangerous for anyone relying on the previous behavior

I'm fine with deprecating it and providing a new `replaceOne` method though

alechenninger commented:
Why keep the old behavior? Literally all it did was throw an
IllegalArgumentException.

On Sun, Aug 5, 2018, 2:35 PM Ben McCann <notifications@github.com> wrote:

> *@benmccann* commented on this pull request.
> ------------------------------
>
> In src/main/java/org/mongojack/JacksonMongoCollection.java
> <https://github.com/mongojack/mongojack/pull/171#discussion_r207745025>:
>
> >      public UpdateResult update(DBQuery.Query query, T object, boolean upsert, WriteConcern concern) throws MongoException, MongoWriteException,
>              MongoWriteConcernException {
> -        if (concern != null) {
> -            return mongoCollection.withWriteConcern(concern).updateOne(serializeQuery(query), convertToDocument(object), new UpdateOptions().upsert(
> -                    upsert));
> -        } else {
> -            return mongoCollection.updateOne(serializeQuery(query), convertToDocument(object), new UpdateOptions().upsert(upsert));
> -        }
> -
> +        return replaceOne(query, object, upsert, concern);
>
> Let's not change the implementation of this method. It would be dangerous
> for anyone relying on the previous behavior
>
> I'm fine with deprecating it and providing a new replaceOne method though
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mongojack/mongojack/pull/171#pullrequestreview-143419443>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADVab43CNmaRkY_rEo3X1HFmuw54HIBZks5uNzsNgaJpZM4T0QWh>
> .
>

benmccann commented:
Ah, sorry I hadn't read your previous comment. In that case, with the additional explanation, I think I actually prefer removing the methods to deprecating them. If all they do is throw an exception then there should be no one calling them, so there is no harm in removing them.

I confirmed that the methods don't work today. When I try to call them I get `java.lang.IllegalArgumentException: Invalid BSON field name _id`

alechenninger commented:
Thanks so much @benmccann ! Do you know when you'll be cutting another release? That way I can stop using my fork :-).

benmccann commented:
Ok. Will cut a release now. It usually takes an hour to show up on Maven Central
